### PR TITLE
[needs-review] Replace Core Concepts articles for Getting Started, Pages, Widgets, and Pieces

### PR DESCRIPTION
**Pull request by stephenkostas**
This PR has the Pages, Widgets, and Pieces rewrites. Because the rewrite process is incomplete, most of the original articles are still intact and linked to for more information.
PR also includes updated redirects, and the complete project with the code for the tutorial (may want to remove that before publishing)
@localghost443

### CODE REVIEW

File: advanced-topics/advanced-pieces-topics/suppressing-pieces-from-a-listing.md [hunk 1]
```diff
@@ -1,41 +0,0 @@
----
```
Comment (boutell) [inline advanced-topics/advanced-pieces-topics/suppressing-pieces-from-a-listing.md]: Why was this file deleted?

File: code/example-project/LICENSE [hunk 2]
```diff
@@ -0,0 +1,7 @@
+Copyright (c) 2013 P'unk Avenue LLC
```
Comment (boutell) [inline code/example-project/LICENSE]: Copyright (c) 2020 Apostrophe Technologies

File: code/example-project/README.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
+# Apostrophe Boilerplate v2.x
```
Comment (boutell) [inline code/example-project/README.md]: Update to suit the new project

File: code/example-project/app.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
+var path = require('path');
+
+var apos = require('apostrophe')({
+  shortName: 'final-test',
```
Comment (boutell) [inline code/example-project/app.js]: If we're going to ship the example project code, shortName should be `example-project`

File: code/example-project/app.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    // If a template is not found somewhere else, serve it from the top-level
+    // `views/` folder of the project
+
+    'apostrophe-templates': { viewsFolderFallback: path.join(__dirname, 'views') },
+
+    'link-widgets': {},
+    'foods': {},
+    'foods-widgets': {
+        extend: 'apostrophe-pieces-widgets'
+         },
```
Comment (boutell) [inline code/example-project/app.js]: Formatting is inconsistent. Not sure if you intend to ship this, you mentioned being unsure about it.

File: code/example-project/lib/modules/apostrophe-express/index.js [hunk 6]
```diff
@@ -0,0 +1,6 @@
+module.exports = {
+  session: {
+    // If this still says `undefined`, set a real secret!
+    secret: '4667320c1d460e64'
```
Comment (boutell) [inline code/example-project/lib/modules/apostrophe-express/index.js]: I'd set this to something like CHANGEME if this ships.

File: code/example-project/lib/modules/foods/index.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
+module.exports = {
+    extend: 'apostrophe-pieces',
+    name: 'foods',
+    label: 'Food',
+    pluralLabel: 'Food',
+    addFields: [
+    {
+            name: 'title',
```
Comment (boutell) [inline code/example-project/lib/modules/foods/index.js]: Inconsistent formatting. This project should have an eslintrc file, taken from one of our client projects, with the appropriate dependencies in package.json, and then "npx eslint . --fix" will fix most of this for you.

File: tutorials/core-concepts/pages/pages-example.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+          %}
+            current
+          {% endif %}
+        "><a href="{{ tab._url }}">{{ tab.title }}</a></li>
+      {% endfor %}
+    </ul>
+  ```
+3. Add the same code to `lib/modules/apostrophe-pages/views/pages/home.html` to add it to the Home page as well.
+
+This adds a simple, unstyled navigation that you can use to easily navigate across all your site.
```
Comment (boutell) [inline tutorials/core-concepts/pages/pages-example.md]: It adds navigation just to all of the children of the homepage. That should be made clear. If the site has nested pages there won't be any way to reach them. That doesn't mean you can't have just this one example in this article, just make it clear what it does and doesn't cover, and what you might learn by reading that other article.

File: tutorials/core-concepts/pages/pages.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+The page template determines the page’s layout and what kind of content can be added to a page. You can create a new page template in two steps:
+
+1. Create a new template in `lib/modules/apostrophe-pages/views/pages` which contains the code to define your page layout.
+
+![](/.gitbook/assets/pages-directory-tree.png)
+
+2. Register the template in `app.js`
+
+{% hint style="info" %}
+Apostrophe uses the [Nunjucks templating language](https://mozilla.github.io/nunjucks/). Nunjucks uses a mix of standard HTML and CSS with Nunjucks specific tags. You can learn more about using Nunjucks with Apostrophe in the  [Nunjucks Helper Functions](/tutorials/core-concepts/working-with-templates/nunjucks-helper-functions.md) and [Nunjucks Filters](/tutorials/core-concepts/working-with-templates/nunjucks-filters.md) references.
```
Comment (boutell) [inline tutorials/core-concepts/pages/pages.md]: This makes it sound like Nunjucks has an opinion about HTML and CSS. It doesn't, it's just a template language for outputting text. I would say something like "Nunjucks allows you to write your HTML markup without repeating yourself, and makes it easy to insert dynamic content like page titles and editable content areas."

File: tutorials/core-concepts/pages/pages.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+![](/.gitbook/assets/pages-template-options.png)
+
+After you create pages, any edits to a templates will be propagated to all pages created from that template.
+
+## The page tree
+
+All the pages you create are connected through the page tree. There are two types of relationships between pages: **child** and **ancestor**. Within the page tree, Home is the root. \
+
+*   Every page you create is a **child** of Home.
```
Comment (boutell) [inline tutorials/core-concepts/pages/pages.md]: Every page you create is a descendant of the Home. page. Pages can be nested, so a page might be the child of a page other than the home page.

File: tutorials/core-concepts/pages/pages.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+*   Home is the **ancestor** of every page.
+*   When you create a new page nested under another page, those are **child** pages, and the page they’re nested under are **ancestors**. \
```
Comment (boutell) [inline tutorials/core-concepts/pages/pages.md]: Grammar. "And the pages they are nested beneath are ancestors."

File: tutorials/core-concepts/pieces/README.md [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+title: "Reusable Content With Pieces"
+layout: reference
+children:
+  - reusable-content-with-pieces
+  - displaying-pieces-with-widgets
+  - browsing-directory-of-pieces
+  - joins
+---
+
+When you build a website, you could look at it like a puzzle. Each element of the site needs to fit together so you can see the whole picture. In Apostrophe, any kind of structured content is a “piece” and that content becomes pieces of the puzzle that you’re putting together as you build your site. You can use pieces for things ranging from defining a standard format for a blog post to creating a new entity type like “bikes” for a bike shop. Apostrophe uses pieces internally to define different content out of the box – and many other things like users, that you might not think of as “content” – and you can harness that same power to put the pieces of your own puzzle together.
```
Comment (boutell) [inline tutorials/core-concepts/pieces/README.md]: * Tense issue in the first sentence... how about just "when you build a website, look at it like a puzzle"?
* Second sentence: pages are structured content too, and they are not pieces. In Apostrophe, every document is either a page or a piece. Pages have a fixed location in the page tree, while pieces do not. Pieces can be displayed in one or more places on the site depending on their content, which makes them well suited to content types like products, blog posts, events, etc.
* As for ways Apostrophe uses pieces internally, this is a good idea to talk about, but it'll be clearer if we're specific: Apostrophe uses pieces internally to power the media library ("Images" and "Files" on the admin bar), allowing those things to be reused in many places on the site, as well as things you might not think of as content, like users.

File: tutorials/core-concepts/pieces/README.md [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+
+Take a bike shop as an example: you could create individual pages for everything you sell – each type of bike, each type of helmet, and each repair service that you offer – but you’d up with a messy site that would quickly become unmanageable. So instead, you take a step back, and think about the components that you need to build your content, and create a piece type for each one: bikes, helmets, accessories, services, etc.. Each individual item is created as a piece. When you get a new model bike in stock, you create a new piece from the bike piece type, and the associated page and entry is automatically created. In Apostrophe, pieces are the key to designing, building, and maintaining your site.
```
Comment (boutell) [inline tutorials/core-concepts/pieces/README.md]: This makes it sound like pieces are better than pages in some way.

Pages and pieces have distinct use cases. If it lives in only one place on the site and has a parent/child relationship to other pages, it's a page. If it might appear in many places on the site according to your needs, it's a piece.

As for this: "When you get a new model bike in stock, you create a new piece from the bike piece type, and the associated page and entry is automatically created"

An Apostrophe page in the tree is not created, but I know what you mean. So let's try to nail down the difference between a page in the tree, and a permalink for a piece.

I would suggest: "When you get a new model bike in stock, you create a new piece via "Bikes" on the admin bar. Then, if you wish, you can display it anywhere on the site via a widget, and also provide a way to browse all bikes and reach a detail page or "permalink" page for each one with no extra effort via Apostrophe's 'pieces-pages" feature."

File: tutorials/core-concepts/pieces/pieces-example.md [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+
+  ```markup
+    module.exports = {
+      extend: 'apostrophe-pieces',
+      name: 'foods',
+      label: 'Food',
+      pluralLabel: 'Food',
+    
+      addFields: [
+      {
```
Comment (boutell) [inline tutorials/core-concepts/pieces/pieces-example.md]: Should be indented.

File: tutorials/core-concepts/pieces/pieces-example.md [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+  ```
+
+2. Add the following code to the **modules** section of **app.js** to register the piece and the widget: \
+  
+  ```markup
+    modules: { 
+  
+      ... 
+      'foods': {}, 
+      'foods-widgetss': {
```
Comment (boutell) [inline tutorials/core-concepts/pieces/pieces-example.md]: Typo.

File: tutorials/core-concepts/pieces/pieces-example.md [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+
+2. Add the following code to the **modules** section of **app.js** to register the piece and the widget: \
+  
+  ```markup
+    modules: { 
+  
+      ... 
+      'foods': {}, 
+      'foods-widgetss': { 
+      extend: 'apostrophe-pieces-widgets'
```
Comment (boutell) [inline tutorials/core-concepts/pieces/pieces-example.md]: Nesting.

File: tutorials/core-concepts/pieces/pieces-example.md [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+
+![](/.gitbook/assets/pieces-add-food-widget.png) 
+
+```
+Note: in the widget template, you defined a **for loop** which looks through the list of all pieces to display. When you display the pieces in the widget, the selection you make in the dialog box will define the complete list of pieces that the template sees.
+```
+
+![](/.gitbook/assets/pieces-widget-page.png) 
+
+Congratulations! You created a new content type and a widget to display it. However, there are some limitations to using a widget to display this sort of content. One limitation is that everything you do and configure is contained within the instance of the widget placed on a specific page. Another is that the pieces themselves don’t have a “home” that you can universally link to them from. Using pieces pages will solve this.
```
Comment (boutell) [inline tutorials/core-concepts/pieces/pieces-example.md]: This paragraph is very clear, maybe it can be a source to improve the earlier paragraph that was confusing on this point.

Comment (boutell) [reply tutorials/core-concepts/pieces/pieces-example.md]: Another important limitation to be mentioned is that a widget is not a practical way to let users browse a thousand foods. Pieces pages are designed to support that use case through pagination and filtering.

File: tutorials/core-concepts/pieces/pieces-example.md [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+    modules { 
+      ... 
+      'foods-pages': { 
+          extend: 'apostrophe-pieces-pages'
+        }, 
+      ...
+    }
+  ```
+
+2. Then register the pieces pages as a new page type by adding this is the `pages` subsection of `modules`:
```
Comment (boutell) [inline tutorials/core-concepts/pieces/pieces-example.md]: Grammar. Also, `apostrophe-pages` is the module name, not `pages`. Please review this section for correctness.

File: tutorials/core-concepts/pieces/pieces-example.md [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+
+  ```markup
+    {
+      name: 'foods-pages',
+      label: 'Food’'
+    }
+  ```
+
+3. Open `index.html` and add the following code:
```
Comment (boutell) [inline tutorials/core-concepts/pieces/pieces-example.md]: Open it where? Maybe I missed it but I don't see any discussion of the way the folders are laid out here or what `--pages` did for the developer.

File: tutorials/core-concepts/pieces/pieces-example.md [hunk 20]
```diff
@@ -0,0 +1,242 @@
...
+
+3. Name the page “Menu” and click _Save_.
+
+![](/.gitbook/assets/pieces-create-pieces-page.png) 
+
+This creates the piece index page. The `index.html` template generates links for each piece of the piece type.
+
+![](/.gitbook/assets/pieces-menu-page.png) 
+
+The pages for each piece are generated from the `show.html`, so you now have individual pages for each piece.
```
Comment (boutell) [inline tutorials/core-concepts/pieces/pieces-example.md]: I get why you keep calling these "pages," it's a bit exhausting to keep distinguishing between pages in the tree and these "virtual" pages without even having a clear term for it. So how about we just call them that consistently — virtual pages?

File: tutorials/core-concepts/pieces/pieces-example.md [hunk 21]
```diff
@@ -0,0 +1,242 @@
...
+
+![](/.gitbook/assets/pieces-pieces-page.png) 
+
+The individual pieces pages will even integrate back with the widget, since now each piece has its own page at its own URL.
```
Comment (boutell) [inline tutorials/core-concepts/pieces/pieces-example.md]: How about:

The widget will even offer links to the individual virtual pages for the pieces, now that an index page is available to provide a starting point for those URLs.

File: tutorials/core-concepts/pieces/pieces-example.md [hunk 22]
```diff
@@ -0,0 +1,242 @@
...
+
+## Next Steps
+
+From this example you can see how powerful pieces are, but we’ve only just scratched the surface. You can continue reading to try a [different piece example which uses more features, like joins](joins.md), or you can go to the [Advanced Pieces](/tutorials/advanced-development/advanced-pieces-topics) to learn more about concepts like [filters](/tutorials/advanced-development/advanced-pieces-topics/adding-filters.md) and [columns](/tutorials/advanced-development/advanced-pieces-topics/adding-columns.md).
```
Comment (boutell) [inline tutorials/core-concepts/pieces/pieces-example.md]: I don't think that first link to joins is well-motivated. Tell the user what they will get from that example.

File: tutorials/core-concepts/pieces/pieces.md [hunk 23]
```diff
@@ -0,0 +1,64 @@
+---
```
Comment (boutell) [inline tutorials/core-concepts/pieces/pieces.md]: If I come straight to this page from a google search I won't know what a piece is. That should be recapped.

File: tutorials/core-concepts/pieces/pieces.md [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+## How to create a new piece type
+
+You can create a new piece type in two steps:
+
+1. Create a module for the new piece type.
+2. Define the structure, or [schema](/tutorials/schema-guide/schema-guide.md), within the piece module.
+
+### Create a new module for the piece type
+
+To create a module, all you need to do is create a new folder, named for your new piece type, inside of your projects `lib/modules` folder (or use the `apostrophe create-piece` command from the CLI), and then register the module in `app.js`.
```
Comment (boutell) [inline tutorials/core-concepts/pieces/pieces.md]: Or you can use the CLI, as you did earlier. Should discuss both.

File: tutorials/core-concepts/pieces/pieces.md [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+
+![](/.gitbook/assets/pieces-menu.png) 
+
+You can easily add and manage entries through the administration interface that Apostrophe generates for each piece type.
+
+![](/.gitbook/assets/pieces-manage.png) 
+
+## Displaying pieces
+
+Each piece type needs a widget or a piece page to display it. Within the piece pages or widget, you have access to the schema fields from the piece. You can access these in the templates by prepending `piece`. to the name of the field. For example, if you wanted to the title of the piece, you could access it with `piece.title` or if you had a field for the phone number named “phone” you could access it in the template with `piece.phone`.
```
Comment (boutell) [inline tutorials/core-concepts/pieces/pieces.md]: In the templates you must start with `data`, i.e. `data.piece.title`.

File: tutorials/core-concepts/pieces/pieces.md [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+
+### Pieces Widgets
+
+You can [create a widget](/tutorials/core-concepts/widgets) for displaying pieces of a specific type. To do create a piece widget:
+
+1. Create and register the widget.
+
+2. Name the widget using the convention `<piece-name>-widgets`, this will provide you access to piece data in the widget without [further configuration](link)
```
Comment (boutell) [inline tutorials/core-concepts/pieces/pieces.md]: Name the module providing the widget

Comment (boutell) [reply tutorials/core-concepts/pieces/pieces.md]: Dead link?

File: tutorials/core-concepts/pieces/pieces.md [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+
+3. Add your code to the widget
```
Comment (boutell) [inline tutorials/core-concepts/pieces/pieces.md]: Needs detail

File: tutorials/core-concepts/pieces/pieces.md [hunk 28]
```diff
@@ -0,0 +1,64 @@
...
+
+4. Add the widget to a page template in a singleton or area
+
+
+### Pieces Pages
+
+Piece pages enables you to create a page which displays all pieces of a certain type on one page, and also provide individual pages for each piece. The main page is called a **piece index page**, and the individual pages are called **piece show pages.**
```
Comment (boutell) [inline tutorials/core-concepts/pieces/pieces.md]: This duplicates material earlier in the PR? It's clearer than earlier in the PR though. 

I would say "individual virtual pages for each piece" and try to stick with that convention when not seeking to be more precise ("piece show page" is also good).

File: tutorials/core-concepts/pieces/pieces.md [hunk 29]
```diff
@@ -0,0 +1,64 @@
...
+
+As an example, if you were creating a site for a restaurant, the piece index page could be the menu which displays the name of each item and its price, and then the piece show pages could have pictures and nutritional information for each individual item.
+
+To create pieces pages:
+
+1. Create the module and register it as extending **apostrophe-pieces-pages**.
```
Comment (boutell) [inline tutorials/core-concepts/pieces/pieces.md]: Does the new documentation ever address the steps to create a module, as such?

File: tutorials/core-concepts/widgets/widgets-example.md [hunk 30]
```diff
@@ -0,0 +1,135 @@
+---
+title: Example: Creating and Displaying a Simple Widget
+layout: tutorial
+---
+
+# Example: Creating and Displaying a Simple Widget
+
+To demonstrate creating a widget, start simple, and create a simple customizable link widget. Then display it in two locations - first in a repeatable form in an area on the body of the page and then in the footer of the page with as a singleton. d
```
Comment (boutell) [inline tutorials/core-concepts/widgets/widgets-example.md]: Stray letter at end

File: tutorials/core-concepts/widgets/widgets-example.md [hunk 31]
```diff
@@ -0,0 +1,135 @@
...
+  ```
+
+In the page template, the widget is in an “area.” And within the area, and editor can add any number of links.
+
+Next, add a singleton to the footer.
+
+1. Add this `afterMain` section to the bottom of default.html:
+
+  ```markup
+    {% block afterMain %} \
```
Comment (boutell) [inline tutorials/core-concepts/widgets/widgets-example.md]: What's up with the slashes?

File: tutorials/core-concepts/widgets/widgets-example.md [hunk 32]
```diff
@@ -0,0 +1,135 @@
...
+    {% code-tabs %} {% code-tabs-item title="app.js" %}
+    modules: {
+    
+    // ...,
+    'link-widgets': {}
+    
+    }
+  ```
+
+3. Now define your widgets schema in `lib/modules/link-widgets/index.js` in:
```
Comment (boutell) [inline tutorials/core-concepts/widgets/widgets-example.md]: Grammar.

File: tutorials/core-concepts/widgets/widgets.md [hunk 33]
```diff
@@ -0,0 +1,42 @@
...
+
+## Creating a Widget
+
+You create a widget by creating a new module (link). A module is basically anything that is deployed to Apostrophe. It could be a collection of pages, widget, or a new content type.
+
+You can create a widget in just a few steps:
+
+1. Create the widget with the CLI. 
+
+2. Register the widget in app.js.
```
Comment (boutell) [inline tutorials/core-concepts/widgets/widgets.md]: Register a module to provide the widget in app.js.

File: tutorials/core-concepts/widgets/widgets.md [hunk 34]
```diff
@@ -0,0 +1,42 @@
...
+
+You can see the [field reference](/other/field-types.md) for more information on all the options you have available for building widgets.
+
+{% hint style="info" %}
+Warning:  Your module name should not start with `apostrophe`. That's reserved for our official modules.
+{% endhint %}
+
+## Displaying a Widget
+
+In Apostrophe, widgets are displayed on a page using tools called Areas and Singletons. An **Area** is a region of a page, defined in a template, where you can add one or more widgets to display. A **Singleton** displays a single widget in a fixed location. Because Areas and Singletons  are defined in templates which are used to create many pages, generally you’ll want the flexibility of Areas in order to add different widgets with different arrangements to pages using the same template, and use Singletons for specific cases where you need to have one widget in a fixed location.
```
Comment (boutell) [inline tutorials/core-concepts/widgets/widgets.md]: This material is fine, but did this not get covered earlier along with how to use all the standard widgets? Shouldn't it be familiar by this point, something to just link back to?

File: tutorials/getting-started/getting-started.md [hunk 35]
```diff
@@ -0,0 +1,147 @@
+---
+title: Getting Started
+layout: tutorial
+---
+
+# Getting Started
+
+In this guide, we’ll cover quickly installing Apostrophe locally for testing and development and testing. First, you’ll need to get the various dependencies installed:
+
+1. Homebrew (Linux and MacOS)
```
Comment (boutell) [inline tutorials/getting-started/getting-started.md]: Linux does not use Homebrew. That's MacOS only. On Linux developers would use the package manager of the Linux distribution they chose.

File: tutorials/getting-started/getting-started.md [hunk 36]
```diff
@@ -0,0 +1,147 @@
...
+2. MongoDB
+3. node.js
```
Comment (boutell) [inline tutorials/getting-started/getting-started.md]: Capitalize the N when using its full name.

File: tutorials/getting-started/getting-started.md [hunk 37]
```diff
@@ -0,0 +1,147 @@
...
+        npm install apostrophe-cli -g
+    ```
+
+2. Next, make sure you’re in the directory where you want your Apostrophe projects located and create your project:
+
+    ```bash
+        apostrophe create-project <your-project-name> && cd <your-project-name> && npm i
+    ```
+
+Running `npm-install` will install Apostrophe with the remaining dependencies. Before you can run Apostrophe you’ll want to create an administrative user.
```
Comment (boutell) [inline tutorials/getting-started/getting-started.md]: This is redundant to the `npm i` on the previous line and the hyphen is incorrect.

File: tutorials/getting-started/old/setting-up-your-environment.md [hunk 38]
```diff
@@ -0,0 +1,148 @@
+---
```
Comment (boutell) [inline tutorials/getting-started/old/setting-up-your-environment.md]: Why these changes to something in "/old" that appear to duplicate the new getting started article?

**Comment by abea** [comment]
I think this is all deprecated at this point. @localghost443 Let me know if you think otherwise.
